cartographer_ros/cartographer_ros/msg_conversion.cc

```diff
@@ -404,10 +404,20 @@ std::unique_ptr<nav_msgs::OccupancyGrid> CreateOccupancyGridMsg(
       const uint32_t packed = pixel_data[y * width + x];
       const unsigned char color = packed >> 16;
       const unsigned char observed = packed >> 8;
-      const int value =
-          observed == 0
-              ? -1
-              : ::cartographer::common::RoundToInt((1. - color / 255.) * 100.);
+
+      int value = -1;
+      if (observed != 0) {
+        const auto cell =
+            ::cartographer::common::RoundToInt((1. - color / 255.) * 100.);
+        if (trinary_interpretation) {
```

Comment:
This is deviating from documentation of this message: http://docs.ros.org/api/nav_msgs/html/msg/OccupancyGrid.html

Did you consider improving the nav stack to handle this properly?

Reply:
While it is correct that deviates from the documentation of OccupancyGrid, `map_server` as a standard in ROS uses that same message for publishing in this way.
It's true also that nav stack could be improved to properly work with the current approach that Cartographer has, but I want to bring an option that would allow ANY package developed to work with `map_server` (again, as a standard), to properly work with Cartographer too. As I mentioned this is the case of autonomous exploration packages as `explore-lite` and `frontier-exploration`. There are a lot of issues opened on the GitHub of those packages asking to work with Cartographer, and the same might be happening with others packages that  I don't know.

Reply:
This sounds unconvincing. It seems you agree that the current implementation is correct. Also `map_server` which you mention seems to use the message as documented. http://wiki.ros.org/map_server even mentions `scale` and `raw` modes which would also break the other packages you are interested in it would seem at first glance. Is this not the case? I can only encourage you to investigate improvements in these packages. Not foremost to support the approach that Cartographer has, but to the support the OccupancyGrid message as it is documented.